### PR TITLE
gh-106751: Optimize SelectSelector.select() for many iteration case

### LINKED ISSUE #106751: Optimize XXXSelector for many iterations of the event loop
split out from https://github.com/python/cpython/pull/106555#issuecomment-1632025959

The current `EpollSelector` can be sped up a bit. This makes quite a difference when there are 100000+ iterations of the event loop per minute (the use case being receiving bluetooth data from multiple sources) since selectors have to run every iteration.

original: 11.831302762031555
new: 9.579423972172663

```
import timeit
import math
import select
import os
from selectors import EpollSelector, EVENT_WRITE, EVENT_READ


class OriginalEpollSelector(EpollSelector):
    def select(self, timeout=None):
        if timeout is None:
            timeout = -1
        elif timeout <= 0:
            timeout = 0
        else:
            # epoll_wait() has a resolution of 1 millisecond, round away
            # from zero to wait *at least* timeout seconds.
            timeout = math.ceil(timeout * 1e3) * 1e-3
        # epoll_wait() expects `maxevents` to be greater than zero;
        # we want to make sure that `select()` can be called when no
        # FD is registered.
        max_ev = max(len(self._fd_to_key), 1)
        ready = []
        try:
            fd_event_list = self._selector.poll(timeout, max_ev)
        except InterruptedError:
            return ready
        for fd, event in fd_event_list:
            events = 0
            if event & ~select.EPOLLIN:
                events |= EVENT_WRITE
            if event & ~select.EPOLLOUT:
                events |= EVENT_READ

            key = self._key_from_fd(fd)
            if key:
                ready.append((key, events & key.events))
        return ready


NOT_EPOLLIN = ~select.EPOLLIN
NOT_EPOLLOUT = ~select.EPOLLOUT

class NewEpollSelector(EpollSelector):
    def select(self, timeout=None):
        if timeout is None:
            timeout = -1
        elif timeout <= 0:
            timeout = 0
        else:
            # epoll_wait() has a resolution of 1 millisecond, round away
            # from zero to wait *at least* timeout seconds.
            timeout = math.ceil(timeout * 1e3) * 1e-3
        # epoll_wait() expects `maxevents` to be greater than zero;
        # we want to make sure that `select()` can be called when no
        # FD is registered.
        max_ev = len(self._fd_to_key) or 1
        ready = []
        try:
            fd_event_list = self._selector.poll(timeout, max_ev)
        except InterruptedError:
            return ready
        
        fd_to_key = self._fd_to_key
        for fd, event in fd_event_list:
            key = fd_to_key.get(fd)
            if key:
                ready.append(
                    (
                        key,
                        (
                            (event & NOT_EPOLLIN and EVENT_WRITE)
                            | (event & NOT_EPOLLOUT and EVENT_READ)
                        )
                        & key.events,
                    )
                )
        return ready


original_epoll = OriginalEpollSelector()
new_epoll = NewEpollSelector()


for _ in range(512):
    r, w = os.pipe()
    os.write(w, b"a")
    original_epoll.register(r, EVENT_READ)
    new_epoll.register(r, EVENT_READ)


original_time = timeit.timeit(
    "selector.select()",
    number=100000,
    globals={"selector": original_epoll},
)
new_time = timeit.timeit(
    "selector.select()",
    number=100000,
    globals={"selector": new_epoll},
)

print("original: %s" % original_time)
print("new: %s" % new_time)
```

### Linked PRs
* gh-106754
* gh-106864
* gh-106879
* gh-106884

### PR DESCRIPTION
**Pull request by corona10**
closes: gh-106751

### CODE REVIEW

**Comment by corona10** [comment]
Similar technique can be adopted.

## Benchmark
```python
import timeit
import os
from selectors import SelectSelector, EVENT_WRITE, EVENT_READ

kqueue = SelectSelector()
for _ in range(100):
    r, w = os.pipe()
    os.write(w, b"a")
    kqueue.register(r, EVENT_READ)

elapsed = timeit.timeit(
    "selector.select()",
    number=100000,
    globals={"selector": kqueue},
)

print(f"elapsed: {elapsed}")


```
## Result
AS-IS: elapsed -> 2.051669834000222
TO-BE: elapsed -> 1.9753025420000085 (1.04x faster)

**Comment by corona10** [comment]
cc @eendebakpt

**Review by methane — APPROVED** [review]

File: Lib/selectors.py [hunk 1]
```diff
@@ -314,17 +314,15 @@ def select(self, timeout=None):
...
-        w = set(w)
-        for fd in r | w:
-            events = 0
-            if fd in r:
-                events |= EVENT_READ
-            if fd in w:
-                events |= EVENT_WRITE
-
-            key = self._fd_to_key.get(fd)
+        r = frozenset(r)
```
Comment (eendebakpt) [inline Lib/selectors.py:317]: Note: on my system `frozenset` is slightly slower than `set`, but the difference is very small

File: Lib/selectors.py [hunk 2]
```diff
@@ -314,17 +314,15 @@ def select(self, timeout=None):
...
+        w = frozenset(w)
+        rw = r | w
```
Comment (eendebakpt) [inline Lib/selectors.py:319]: The `rw` is only used once. You assigned `r | w` to a variable for readability?

**Pull request merged**